src/me.rs

```diff
@@ -18,54 +19,223 @@ use crate::mc::MotionVector;
+struct MVSearchResult {
+  mv: MotionVector,
+  cost: u64,
+}
+
+#[derive(Debug, Copy, Clone)]
+struct FullpelSearchResult {
   mv: MotionVector,
   cost: u64,
+  sad: u32,
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+enum MVSamplingMode {
+  INIT,
+  CORNER { right: bool, bottom: bool },
+}
+
+#[hawktracer(estimate_tile_motion)]
+pub fn estimate_tile_motion<T: Pixel>(
+  fi: &FrameInvariants<T>, ts: &mut TileStateMut<'_, T>,
+  inter_cfg: &InterConfig,
+) {
+  let init_size = MIB_SIZE_LOG2;
+  for mv_size_log2 in (2..=init_size).rev() {
+    let init = mv_size_log2 == init_size;
+
+    // Choose subsampling. Pass one is quarter res and pass two is at half res.
+    let ssdec = match init_size - mv_size_log2 {
+      0 => 2,
+      1 => 1,
+      _ => 0,
+    };
+
+    for sby in 0..ts.sb_height {
+      for sbx in 0..ts.sb_width {
+        let mut tested_frames_flags = 0;
+        for &ref_frame in inter_cfg.allowed_ref_frames() {
+          let frame_flag = 1 << fi.ref_frames[ref_frame.to_index()];
+          if tested_frames_flags & frame_flag == frame_flag {
+            continue;
+          }
+          tested_frames_flags |= frame_flag;
+
+          estimate_sb_motion(
+            fi,
+            ts,
+            ref_frame,
+            mv_size_log2,
+            TileSuperBlockOffset(SuperBlockOffset { x: sbx, y: sby })
+              .block_offset(0, 0),
+            init,
+            ssdec,
+          );
+        }
+      }
+    }
+  }
+}
+
+fn estimate_sb_motion<T: Pixel>(
+  fi: &FrameInvariants<T>, ts: &mut TileStateMut<'_, T>, ref_frame: RefType,
+  mut mv_size_log2: usize, tile_bo: TileBlockOffset, init: bool, ssdec: u8,
+) {
+  let size_mi = MIB_SIZE;
+  let h_in_b: usize = size_mi.min(ts.mi_height - tile_bo.0.y);
+  let w_in_b: usize = size_mi.min(ts.mi_width - tile_bo.0.x);
+  let mut edge_mode = false;
+
+  // Check for edge of the frame where size_mi doesn't fit neatly.
+  let horz_edge = h_in_b != size_mi;
+  let vert_edge = w_in_b != size_mi;
+
+  loop {
+    let mv_size = 1 << mv_size_log2;
+    let bsize = BlockSize::from_width_and_height(
+      mv_size << MI_SIZE_LOG2,
+      mv_size << MI_SIZE_LOG2,
+    );
+
+    // Clip subsampling to ensure sad is computed in chunk of 4x4 and up.
+    let ssdec = ssdec.min(mv_size_log2 as u8);
+
+    // Skip rows that have already been processed.
+    let y_start = if !(edge_mode && horz_edge) {
+      0
+    } else {
+      h_in_b & (!0 << mv_size_log2 << 1)
+    };
+
+    // Process in blocks, excluding regions that cannot fit evenly into a block.
+    // Will either process starting at the first block or only on the edges.
+    for y in (
+      // Don't skip the vertical edge.
+      if vert_edge { 0 } else { y_start as isize }
+        ..=h_in_b as isize - mv_size as isize
+    )
+      .step_by(mv_size)
+    {
+      // Process unprocessed columns and the horizontal edge.
+      let x_start = if !(edge_mode && vert_edge) || y as usize >= y_start {
```

Comment:
> \<derf\> maseck: My only remaining question is if you really meant `> y_start` or `>= y_start`.

Reply:
Yeah, I meant >= instead of > so I changed it.